Генератор значений для тестов

In [21]:
import neftpy.upvt_oil as pvto
import neftpy.uconst as uc
import neftpy.upvt_np_vect as pvt
import neftpy.upvt_gas as pvtg

import numpy as np

In [2]:
"""
Импорт унифлока зависит от корректных путей к надстройке
Путь задается вручную в явном виде (так проще всего)
"""

unifloc_path = r'D:\unifloc\unifloc_vba'
unifloc_xlam = unifloc_path +r'\UniflocVBA_7.xlam'
print('Путь к надстройке Unifloc VBA -- '+ unifloc_xlam)

import sys
sys.path.insert(0, unifloc_path) # добавим в путь поиска пакетов python папку где находится репозиторий Unifloc VBA
import unifloc_vba_python_api.python_api as unifloc # импортируем python_api для Unifloc VBA

# создаем объект имеющий доступ к расчетам Unifloc VBA
unf = unifloc.API(unifloc_xlam)

print('объект unf обеспечивает доступ к API Unifloc VBA')

# кроме подключения к API подключимся низкоуровнево, чтобы видеть все функции
import xlwings as xw
book = xw.Book(unifloc_xlam)

Путь к надстройке Unifloc VBA -- D:\unifloc\unifloc_vba\UniflocVBA_7.xlam
объект unf обеспечивает доступ к API Unifloc VBA


# unf_pb_Standing_MPaa

In [3]:
# аналог pvto.unf_pb_Standing_MPaa в унифлок vba unf_pvt_pb_Standing_MPa

unf_pvt_pb_Standing_MPa = book.macro("unf_pvt_pb_Standing_MPa")

In [4]:
rsb_m3m3 = 100
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 350
print(f'neftpy = {pvto.unf_pb_Standing_MPaa(t_K, rsb_m3m3, gamma_oil, gamma_gas)}') 
print(f'unifloc vba =  {unf_pvt_pb_Standing_MPa(rsb_m3m3, gamma_gas, t_K, gamma_oil)}') 
print(f'neftpy плотность воздуха {uc.RHO_AIR_kgm3}')

#   разница значений возникает из за разных оценок плотности воздуха
#   

neftpy = 19.2247953146684
unifloc vba =  20.170210695316566
neftpy плотность воздуха 1.2041512897703845


разница значений при расчете neftpy и unifloc vba возникает из за разных оценок плотности воздуха

* unifloc_vba в расчет неявно сидит в формула $\rho_{air} = 1.225$  (соответствут плотности при 15С)
* в neftpy явно выделена константа где $\rho_{air} =1.20$  (соответствут плотности при 20С)

In [5]:

# векторные значения импортированы из расчета unifloc vba
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 353
unf_vba_rsb = [1,60.8,120.6,180.4,240.2,300]

rsb_m3m3 = np.array(unf_vba_rsb)    

pvt.unf_pb_Standing_MPaa(t_K, rsb_m3m3, gamma_oil, gamma_gas)

array([ 0.42998783, 12.86606364, 22.71558836, 31.73087492, 40.24216312,
       48.39676869])

# unf_rs_Standing_m3m3

In [6]:
# аналог pvto.unf_rs_Standing_m3m3 в унифлок vba unf_pvt_rs_Standing_m3m3

unf_pvt_rs_Standing_m3m3 = book.macro("unf_pvt_rs_Standing_m3m3")

In [7]:
p_MPaa = 10
pb_MPaa = 15
rsb = 200
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 350
pvto.unf_rs_Standing_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)
#                        122.74847910146916, 
#                        delta=0.0001)

array(122.7484791)

In [8]:
p_MPaa = 10
pb_MPaa = 0
rsb = 200
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 350
pvto.unf_rs_Standing_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)
#                        122.74847910146916, 
#                        delta=0.0001)


print(f'neftpy = {pvto.unf_rs_Standing_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)}') 
a = 1.
print(f'unifloc vba =  {unf_pvt_rs_Standing_m3m3(p_MPaa, gamma_gas, t_K, gamma_oil, a)}') 
print(a)

neftpy = 45.24604018198833
unifloc vba =  42.70500034810285
1.0


тут тоже разница расчетов из за разной плотности воздуха и соответственно разных значений gamma_gas

In [9]:

gamma_oil = 0.86
gamma_gas = 0.6
t_K = 353
unf_vba_res = [2.63345652886037,15.6762859290458,31.2325939746016,48.2819334435268,66.436537812051,85.4819163871848]
unf_vba_p_MPaa = [1,4.4,7.8,11.2,14.6,18]
pb_MPaa = 0
rsb = 120


p_MPaa = np.array(unf_vba_p_MPaa)
res = np.array(unf_vba_res)

pvt.unf_rs_Standing_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)

array([ 2.79015288, 16.60905883, 33.09100083, 51.15481285, 70.38965542,
       90.56827519])

In [10]:
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 353
unf_vba_res = [2.63345652886037,15.6762859290458,31.2325939746016,48.2819334435268,66.436537812051,85.4819163871848]
unf_vba_p_MPaa = [1,4.4,7.8,11.2,14.6,18]
pb_MPaa = 0
rsb = 120


p_MPaa = np.array(unf_vba_p_MPaa)
res = np.array(unf_vba_res)

pvt.unf_rs_Standing_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas, calc_drs_dp=True)

array([3.35934407, 4.54484246, 5.10789295, 5.49914238, 5.80473597,
       6.0580113 ])

In [11]:
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 353
unf_vba_res = [2.63345652886037,15.6762859290458,31.2325939746016,48.2819334435268,66.436537812051,85.4819163871848]
unf_vba_p_MPaa = [1,4.4,7.8,11.2,14.6,18]
pb_MPaa = 10
rsb = 120


p_MPaa = np.array(unf_vba_p_MPaa)
res = np.array(unf_vba_res)

pvt.unf_rs_Standing_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas, calc_drs_dp=True)

array([ 7.50207231, 10.14952206, 11.40692392,  0.        ,  0.        ,
        0.        ])

# unf_rs_Velarde_m3m3

In [12]:
# аналог pvto.unf_rs_Velarde_m3m3 в унифлок vba unf_pvt_rs_Velarde_m3m3

unf_pvt_rs_Velarde_m3m3 = book.macro("unf_pvt_rs_Velarde_m3m3")

In [14]:
p_MPaa = 10
pb_MPaa = 15
rsb = 250
gamma_oil = 0.86
gamma_gas = 0.6
t_K = 350
pvto.unf_rs_Velarde_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)
#                        170.25302712587356,
#                        delta=0.0001)


print(f'neftpy = {pvto.unf_rs_Velarde_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)}') 
a = 1.
print(f'unifloc vba =  {unf_pvt_rs_Velarde_m3m3(p_MPaa, pb_MPaa, gamma_gas, t_K, gamma_oil, rsb, a)}') 
print(a)

neftpy = 170.2531379850611
unifloc vba =  170.13943535146635
1.0


In [16]:
p_MPaa = 10
pb_MPaa = 15
rsb = 250
gamma_oil = 0.86
gamma_gas = 0.7
t_K = 350
pvto._unf_rs_Velarde_m3m3_(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)
#                        170.25302712587356, 
#                        delta=0.0001)

np.float64(179.5022473749806)

In [17]:

gamma_oil = 0.86
gamma_gas = 0.6
t_K = 353
pb_MPaa = 12.8300024673082
rsb = 120
unf_vba_res = [13.769156001658,47.1724892015713,76.4710407632712,105.644331604199,120,120]
unf_vba_p_MPaa = [1,4.4,7.8,11.2,14.6,18]


p_MPaa = np.array(unf_vba_p_MPaa)        
res = np.array(unf_vba_res)

# получается сходимость на уровне 1 знак после запятой с unifloc vba, что для газосодержания в принципе допустимо
# причины расхождени видимо в количестве учитываемых знаков после запятой в константах или в определении стандартных величин
# принты далее могут показать расхождения по отдельным значениям
#print(unf_vba_res)
#print(pvtovect.unf_rs_Velarde_m3m3(p_MPaa, pb_MPaa, rsb, gamma_oil, gamma_gas, t_K))

pvt.unf_rs_Velarde_m3m3(p_MPaa, t_K, pb_MPaa, rsb, gamma_oil, gamma_gas)
 #                           res, 
 #                           rtol=0.01))

array([ 13.79399931,  47.23193583,  76.53074165, 105.67222853,
       120.        , 120.        ])

# unf_bo_below_pb_m3m3

In [18]:
# аналог pvto.unf_bo_below_pb_m3m3 в унифлок vba unf_pvt_FVF_McCain_m3m3

unf_pvt_FVF_McCain_m3m3 = book.macro("unf_pvt_FVF_McCain_m3m3")

In [20]:
density_oilsto_kgm3 = 800
rs_m3m3 = 200
density_oil_kgm3 = 820
gamma_gas = 0.6
pvto.unf_bo_below_pb_m3m3(density_oilsto_kgm3, density_oil_kgm3,  rs_m3m3,  gamma_gas)
#                        1.1542114715227887, 
#                        delta=0.001
#                        )

print(f'neftpy = {pvto.unf_bo_below_pb_m3m3(density_oilsto_kgm3, density_oil_kgm3,  rs_m3m3,  gamma_gas)}') 

print(f'unifloc vba =  {unf_pvt_FVF_McCain_m3m3(rs_m3m3,  gamma_gas, density_oilsto_kgm3, density_oil_kgm3)}') 


neftpy = 1.1518270180151782
unifloc vba =  1.1543175609756098


газ

In [22]:
gamma_gas = 0.6
pvtg.unf_pseudocritical_Standing_p_MPa_t_K(gamma_gas)
#                               (5.09893164741181, 198.0708725589674),
#                               rtol=0.0001))

(4.567119999999999, 198.8016)